_posts/2018-08-02-read-only-functions.md

```diff
@@ -0,0 +1,122 @@
+---
+
+In this post I'll highlight one of the ways we're making OpenFaaS a more secure environment for your production functions and workloads.
+
+## What is a read-only function?
+
+Any function in OpenFaaS ia always built into an immutable Docker image, so that you get exactly the same results wherever you want to run your code. A read-only function runs in a Docker image with enhanced security so that users cannot make changes to the underlying file-system in the container or the code of the function either.
+
+## How does it work?
+
+In both Kubernetes and Docker Swarm, we enabled a feature in the built-in security context to make the root file-system `/` read-only. Before we applied this patch to the OpenFaaS providers we also changed all the OpenFaaS templates to run as non-root users to prevent tampering with system files, this change goes one step further and prevents a function from tampering with its own code.
+
+To see this in action with a regular Docker container try the following:
+
+```bash
+docker run --rm -ti --read-only alpine:3.7
+
+# apk add --no-cache curl
+ERROR: Unable to lock database: Read-only file system
+ERROR: Failed to open apk database: Read-only file system
+```
+
+### How do I try it?
+
+This feature is released and available with the latest versions of OpenFaaS including the CLI (0.6.17), Kubernetes faas-netes & operator (0.8.4/0.6.0) and Swarm (0.4.0) support.
+
+* Create a function
+
+```bash
+faas-cli new --lang node overwrite-me --prefix=alexellis2
+mv overwrite-me.yml stack.yml
+echo "alexellis" > ./overwrite-me/valid_user.txt
+```
+
+* Update the function code: `./overwrite-me/handler.js`:
+
+```js
+"use strict"
+
+const fs = require('fs');
+
+module.exports = (context, callback) => {
+    if(process.env["Http_Query"] == "update=1") {
+      fs.writeFile("/home/app/function/valid_user.txt", context, (err) => {
+        return callback(err, {action: "update", value: context});
+      });
+    } else {
+        fs.readFile("/home/app/function/valid_user.txt", "utf8", (err, val) => {
+            return callback(err, {action: "get", value: val});
+        });
+    }
+}
+
+```
+
+* Build/push/deploy
+
+```
+faas-cli build \
+ && faas-cli push \
+ && faas-cli deploy
+```
+
+You can now deploy the function and query the valid user, then try to update it.
+
+```
+echo | faas-cli invoke overwrite-me
+{"action":"get","value":"alexellis\n"}
+
+echo stefanprodan | faas-cli invoke overwrite-me --query update=1
+{"action":"update","value":"stefanprodan\n"}
+
+echo | faas-cli invoke overwrite-me
+{"action":"get","value":"stefanprodan\n"}
+```
+
+Now let's apply a read-only root file-system and see what that looks like:
+
+In the function's stack file edit the following:
+
+```yaml
+functions:
+  overwrite-me:
+    lang: node
+    handler: ./overwrite-me
+    image: alexellis2/overwrite-me:latest
+    readonly_root_filesystem: true
+```
+
+```
+faas-cli deploy
+
+echo stefanprodan | faas-cli invoke overwrite-me 
```

Comment:
This doesn't exercise the same piece of code as the earlier example?  To demonstrate the point shouldn't we run the same commands as above?

Reply:
You're right.. I used the wrong command.